jwst/dq_init/tests/test_dq_init.py

```diff
@@ -0,0 +1,340 @@
+
+    # create raw input data for step
+    dm_ramp = make_rawramp(instrument, nints, ngroups, ysize, xsize, xstart, ystart, exp_type)
+
+    # create a MaskModel for the dq input mask
+    dq, dq_def = make_maskmodel(ysize, xsize)
+
+    # edit reference file with known bad pixel values
+    dq[100, 100] = 2   # Dead pixel
+    dq[200, 100] = 4   # Hot pixel
+    dq[300, 100] = 8   # Unreliable_slope
+    dq[400, 100] = 16  # RC
+    dq[500, 100] = 1   # Do_not_use
+    dq[100, 200] = 3   # Dead pixel + do not use
+    dq[200, 200] = 5   # Hot pixel + do not use
+    dq[300, 200] = 9   # Unreliable slope + do not use
+    dq[400, 200] = 17  # RC + do not use
+
+    # write mask model
+    ref_data = MaskModel(dq=dq, dq_def=dq_def)
+    ref_data.meta.instrument.name = instrument
+    ref_data.meta.subarray.xstart = xstart
+    ref_data.meta.subarray.xsize = xsize
+    ref_data.meta.subarray.ystart = ystart
+    ref_data.meta.subarray.ysize = ysize
+
+    # run do_dqinit
+    outfile = do_dqinit(dm_ramp, ref_data)
+    print(outfile.err.shape)
+
+    if instrument == "FGS":
+        dqdata = outfile.dq
+    else:
+        dqdata = outfile.pixeldq
+
+    # assert that the pixels read back in match the mapping from ref data to science data
+    assert(dqdata[100, 100] == dqflags.pixel['DEAD'])
+    assert(dqdata[200, 100] == dqflags.pixel['HOT'])
+    assert(dqdata[300, 100] == dqflags.pixel['UNRELIABLE_SLOPE'])
+    assert(dqdata[400, 100] == dqflags.pixel['RC'])
+    assert(dqdata[500, 100] == dqflags.pixel['DO_NOT_USE'])
+    assert(dqdata[100, 200] == 1025)
+    assert(dqdata[200, 200] == 2049)
+    assert(dqdata[300, 200] == 16777217)
+    assert(dqdata[400, 200] == 16385)
+
+
+def test_groupdq():
+    # Check that GROUPDQ extension is added to the data and all values are initialized to zero.
+    # size of integration
+    instrument = 'MIRI'
+    nints = 1
+    ngroups = 5
+    xsize = 1032
+    ysize = 1024
+    xstart = 1
+    ystart = 1
+
+    # create raw input data for step
+    dm_ramp = make_rawramp(instrument, nints, ngroups, ysize, xsize, xstart, ystart)
+
+    # create a MaskModel for the dq input mask
+
+    dq, dq_def = make_maskmodel(ysize, xsize)
+    # write mask model
+    ref_data = MaskModel(dq=dq, dq_def=dq_def)
+    ref_data.meta.instrument.name = instrument
+    ref_data.meta.subarray.xstart = xstart
+    ref_data.meta.subarray.xsize = xsize
+    ref_data.meta.subarray.ystart = ystart
+    ref_data.meta.subarray.ysize = ysize
+
+    # run the correction step
+    outfile = do_dqinit(dm_ramp, ref_data)
+
+    # check that GROUPDQ was created and initialized to zero
+    groupdq = outfile.groupdq
+
+    np.testing.assert_array_equal(np.full((1, ngroups, ysize, xsize),
+                                          0,
+                                          dtype=int),
+                                  groupdq,
+                                  err_msg='groupdq not initialized to zero')
+
+
+def test_err():
```

Comment:
I'm not 100% sure this test will succeed, because it's using data models as input, rather than disk files. The `dq_init` step checks for the existence of `pixeldq` and `groupdq` arrays and will create them if they aren't in the data model, but it does not do that for the `err` array. The only place/way the `err` will be initialized if it doesn't exist in the input is when a disk file is loaded into a data model. The check is done at the time the data model is created. So I'm not sure this will work if you're starting with an existing data model (but I could be wrong).

Reply:
This test completes fine on my computer. It does create a 4D ERR array, though with dimensions of [0, 0, 0, 0] rather than matching the dimensions of the data.

Reply:
👍

Reply:
OK, so apparently the creation of `err` and all flavors of `dq` arrays is done when the data model instance is created, rather than when it's populated from a file. That's good.